prelude/toolchains/conan/buckler/conanfile.py

```diff
@@ -0,0 +1,240 @@
+
+        lib_paths = _map_rel_to_root(rootpath, component_info.lib_paths)
+        self.libs = _find_libs(rootpath, lib_paths, component_info.libs)
+        self.system_libs = component_info.system_libs
+
+        self.requires = [_Requirement.parse(req) for req in component_info.requires]
+
+    def generate(self):
+        """Generate Buck2 target definitions for the component."""
+        name = _Requirement(None, self.component_name).to_name(self.package_name)
+        deps = [req.to_label(self.package_name) for req in self.requires]
+        return """\
+
+conan_component(
+    name = {name},
+    defines = {defines},
+    cflags = {cflags},
+    cppflags = {cppflags},
+    include_paths = {include_paths},
+    libs = {libs},
+    static_libs = {static_libs},
+    shared_libs = {shared_libs},
+    system_libs = {system_libs},
+    deps = {deps},
+    package = {package},
+)
+""".format(
+            name = repr(name),
+            defines = repr(self.defines),
+            cflags = repr(self.cflags),
+            cppflags = repr(self.cppflags),
+            include_paths = repr(self.include_paths),
+            libs = repr(list(self.libs.keys())),
+            static_libs = repr({
+                name: sorted(libs.static)
+                for name, libs in self.libs.items()
+                if libs.static
+            }),
+            shared_libs = repr({
+                name: sorted(libs.shared)
+                for name, libs in self.libs.items()
+                if libs.shared
+            }),
+            system_libs = repr(self.system_libs),
+            deps = repr(deps),
+            package = repr(":_package_" + self.package_name))
+
+
+class _BucklerDepCpp(object):
+    """A Conan package."""
+
+    def __init__(self, dep_name, dep_cpp_info, public=False):
+        self.name = dep_name
+        self.public = public
+        self.rootpath = dep_cpp_info.rootpath
+        rootpath = self.rootpath
+        if dep_cpp_info.components:
+            self.components = collections.OrderedDict(
+                (name, _BucklerDepCppComponent(dep_name, component_info))
+                for (name, component_info) in dep_cpp_info.components.items()
+            )
+        else:
+            self.components = collections.OrderedDict([
+                (dep_name, _BucklerDepCppComponent(dep_name, dep_cpp_info))
+            ])
+
+
+    def generate(self):
+        """Generate Buck2 target definitions for the package and its components."""
+        result = """\
+
+conan_dep(
+    name = {name},
+    components = {components},
+    visibility = {visibility},
+)
+""".format(
+            name = repr(self.name),
+            components = repr({
+                name: _Requirement(None, name).to_label(self.name)
+                for name in self.components.keys()
+            }),
+            visibility = repr(["PUBLIC"]) if self.public else repr([]))
+
+        for component in getattr(self, "components", {}).values():
```

Comment:
what is the requirement for getattr here?

Reply:
Good catch, that's a left over from an earlier state where that attribute didn't exist for packages without components. Fixed.